### PR TITLE
Add stable check in IndexLogManager.createLatestStableLog

### PR DESCRIPTION
**Pull request by saucam**
### What changes were proposed in this pull request?

This PR addresses: #56 , added stability state check in ```IndexLogManager.createLatestStableLog```

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Added a new unit test and updated the existing test

### CODE REVIEW

File: src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala [hunk 1]
```diff
@@ -71,7 +71,7 @@ class IndexLogManagerImpl(indexPath: Path) extends IndexLogManager with Logging
       return None
     }
     val contents = FileUtils.readContents(fs, path)
-    Some(LogEntry.fromJson(contents))
+    Try(LogEntry.fromJson(contents)).toOption
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala:74]: If `fromJson` throws a non-fatal exception, this will return `None` instead of throwing an exception, and thus cannot differentiate it with the `!fs.exists` case above. Can we revert this change for now? We can come back to this if required.

File: src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala [hunk 2]
```diff
@@ -111,13 +111,16 @@ class IndexLogManagerImpl(indexPath: Path) extends IndexLogManager with Logging
...
-        logError(s"Failed to create the latest stable log with id = '$id'", ex)
-        false
+    getLog(id) match {
+      case Some(logEntry) if Constants.STABLE_STATES.contains(logEntry.state) =>
+        Try(FileUtil.copy(fs, pathFromId(id), fs, latestStablePath, false, new Configuration))
+          match {
+            case Success(v) => v
+            case Failure(e) =>
+              logError(s"Failed to create the latest stable log with id = '$id'", e)
+               false
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala:121]: nit: indentation.

File: src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala [hunk 3]
```diff
@@ -111,13 +111,16 @@ class IndexLogManagerImpl(indexPath: Path) extends IndexLogManager with Logging
...
+          }
+      case _ => false
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala:123]: I think we need to catch two cases: 1) non stable state log entry and 2) `getLog` returned `None`. Can you also add a logging for each case?

File: src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala [hunk 4]
```diff
@@ -144,11 +144,27 @@ class IndexLogManagerImplTest extends SparkFunSuite with SparkInvolvedSuite with
   test("testUpdateLatestStableLog passes if latestStable.json can be created") {
     val path = new Path(testRoot, UUID.randomUUID().toString)
     val fs = path.getFileSystem(new Configuration)
-    FileUtils.createFile(fs, new Path(path, s"$HYPERSPACE_LOG/0"), "file contents")
+    FileUtils.createFile(
+      fs,
+      new Path(path, s"$HYPERSPACE_LOG/0"),
+      JsonUtils.toJson(getEntry("ACTIVE"))
+    )
     new IndexLogManagerImpl(path).createLatestStableLog(0)
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala:152]: Can you check the return value?

File: src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala [hunk 5]
```diff
@@ -144,11 +144,27 @@ class IndexLogManagerImplTest extends SparkFunSuite with SparkInvolvedSuite with
...
 
+  test("testUpdateLatestStableLog fails if log state is not stable") {
+    val path = new Path(testRoot, UUID.randomUUID().toString)
+    val fs = path.getFileSystem(new Configuration)
+    FileUtils.createFile(
+      fs,
+      new Path(path, s"$HYPERSPACE_LOG/0"),
+      JsonUtils.toJson(getEntry("CANCELLING"))
+    )
+    new IndexLogManagerImpl(path).createLatestStableLog(0)
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala:164]: ditto

File: src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala [hunk 6]
```diff
@@ -144,11 +144,27 @@ class IndexLogManagerImplTest extends SparkFunSuite with SparkInvolvedSuite with
   test("testUpdateLatestStableLog passes if latestStable.json can be created") {
     val path = new Path(testRoot, UUID.randomUUID().toString)
     val fs = path.getFileSystem(new Configuration)
-    FileUtils.createFile(fs, new Path(path, s"$HYPERSPACE_LOG/0"), "file contents")
+    FileUtils.createFile(
+      fs,
+      new Path(path, s"$HYPERSPACE_LOG/0"),
+      JsonUtils.toJson(getEntry("ACTIVE"))
+    )
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala:151]: Can we put this in the previous line? (that's how auto-format is done)

File: src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala [hunk 7]
```diff
@@ -144,11 +144,27 @@ class IndexLogManagerImplTest extends SparkFunSuite with SparkInvolvedSuite with
...
   }
 
+  test("testUpdateLatestStableLog fails if log state is not stable") {
+    val path = new Path(testRoot, UUID.randomUUID().toString)
+    val fs = path.getFileSystem(new Configuration)
+    FileUtils.createFile(
+      fs,
+      new Path(path, s"$HYPERSPACE_LOG/0"),
+      JsonUtils.toJson(getEntry("CANCELLING"))
+    )
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala:163]: ditto

File: src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala (hunk 1 above)
Comment (saucam) [reply src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala:74]: So we want to catch only Non-fatal exceptions here ?

File: src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala (hunk 4 above)
Comment (saucam) [reply src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala:152]: ok

File: src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala (hunk 5 above)
Comment (saucam) [reply src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala:164]: ok

File: src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala (hunk 6 above)
Comment (saucam) [reply src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala:151]: ok

File: src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala (hunk 7 above)
Comment (saucam) [reply src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala:163]: ok

File: src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala (hunk 3 above)
Comment (saucam) [reply src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala:123]: ok

**Comment by saucam** [comment]
@imback82 I have addressed most of your comments! kindly please review again,  Thanks !

File: src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala (hunk 7 above)
Comment (saucam) [reply src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala:171]: done

File: src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala (hunk 6 above)
Comment (saucam) [reply src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala:149]: done

File: src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala [hunk 8]
```diff
@@ -55,6 +55,10 @@ class IndexLogManagerImplTest extends SparkFunSuite with SparkInvolvedSuite with
             Content.Directory("dir2", Seq("1.json", "2.json"), NoOpFingerprint()))))))),
     Map())
 
+  def getEntry(state: String): LogEntry = {
```
Comment (apoorvedave1) [inline src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala:58]: could you please make it private

File: src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala [hunk 9]
```diff
@@ -71,7 +72,13 @@ class IndexLogManagerImpl(indexPath: Path) extends IndexLogManager with Logging
       return None
     }
     val contents = FileUtils.readContents(fs, path)
-    Some(LogEntry.fromJson(contents))
+    Try(LogEntry.fromJson(contents)) match {
+      case Success(s) => Some(s)
+      case Failure(NonFatal(e)) =>
+        logError(s"Error reading LogEntry json from $path, $e")
+        None
+      case Failure(e) => throw e
```
Comment (apoorvedave1) [inline src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala:80]: From this it looks like if `contents` is corrupted, and json conversion fails, it will not throw the error as it is Non-Fatal.
Could you change it to throw in case of all failures? We would not want file to be present but unreadable.

cc @imback82

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala:80]: right, this change should be just reverted as mentioned here: https://github.com/microsoft/hyperspace/pull/92#discussion_r456736210

File: src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala (hunk 8 above)
Comment (saucam) [reply src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala]: ok done

File: src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala (hunk 9 above)
Comment (saucam) [reply src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala]: ok reverted, but it does not look like valid code :
We are returning ```Some``` of something that can throw, but we can essentially never get a ```None``` here

**Comment by saucam** [comment]
Thanks for the comments @apoorvedave1 , fixed your comments, kindly help review again!

cc @imback82

File: src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala [hunk 10]
```diff
@@ -110,16 +110,23 @@ class IndexLogManagerImpl(indexPath: Path) extends IndexLogManager with Logging
...
+      case Some(logEntry) if Constants.STABLE_STATES.contains(logEntry.state) =>
+        Try(FileUtil.copy(fs, pathFromId(id), fs, latestStablePath, false, new Configuration))
+        match {
+          case Success(v) => v
+          case Failure(e) =>
+            logError(s"Failed to create the latest stable log with id = '$id'", e)
+            false
+        }
+      case Some(logEntry) =>
+        logError(s"Found LogEntry with non stable state = ${logEntry.state}")
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala:124]: "for id = '$id'"

File: src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala [hunk 11]
```diff
@@ -110,16 +110,23 @@ class IndexLogManagerImpl(indexPath: Path) extends IndexLogManager with Logging
...
+        false
+      case None =>
+        logError(s"Unable to get LogEntry for id = $id")
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala:127]: quote '$id' to be consistent?

File: src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala [hunk 12]
```diff
@@ -110,16 +110,23 @@ class IndexLogManagerImpl(indexPath: Path) extends IndexLogManager with Logging
...
   }
 
-  override def createLatestStableLog(id: Int): Boolean = {
-    // TODO: make sure log with the id has a stable state.
-    try {
-      FileUtil.copy(fs, pathFromId(id), fs, latestStablePath, false, new Configuration)
-    } catch {
-      case ex: Exception =>
-        logError(s"Failed to create the latest stable log with id = '$id'", ex)
+  override def createLatestStableLog(id: Int): Boolean =
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala:113]: nit: can we put `{` `}` around this `def`?

File: src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala (hunk 6 above)
Comment (imback82) [reply src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala:149]: I meant the parenthesis:
```
FileUtils.createFile(
  fs, new Path(path, s"$HYPERSPACE_LOG/0"), JsonUtils.toJson(getEntry("ACTIVE")))
```

File: src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala (hunk 7 above)
Comment (imback82) [reply src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala:171]: same here

File: src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala [hunk 13]
```diff
@@ -144,11 +144,36 @@ class IndexLogManagerImplTest extends SparkFunSuite with SparkInvolvedSuite with
...
+    assert(result === true)
     assert(fs.exists(new Path(path, s"$HYPERSPACE_LOG/latestStable")))
   }
 
+  test("testUpdateLatestStableLog fails if log state is not stable") {
+    val path = new Path(testRoot, UUID.randomUUID().toString)
+    val fs = path.getFileSystem(new Configuration)
+    FileUtils.createFile(
+      fs, new Path(path, s"$HYPERSPACE_LOG/0"), JsonUtils.toJson(getEntry("CANCELLING"))
+    )
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala:160]: ditto

File: src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala [hunk 14]
```diff
@@ -144,11 +144,36 @@ class IndexLogManagerImplTest extends SparkFunSuite with SparkInvolvedSuite with
...
+
+  test("testUpdateLatestStableLog fails with exception if unable to find a valid log entry") {
+    val path = new Path(testRoot, UUID.randomUUID().toString)
+    val fs = path.getFileSystem(new Configuration)
+    FileUtils.createFile(
+      fs, new Path(path, s"$HYPERSPACE_LOG/0"), "Invalid Log Entry"
+    )
+    assertThrows[com.fasterxml.jackson.core.JsonParseException](
+      new IndexLogManagerImpl(path).createLatestStableLog(0)
+    )
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala:174]: ditto

Comment (saucam) [reply src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala]: done

File: src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala (hunk 13 above)
Comment (saucam) [reply src/test/scala/com/microsoft/hyperspace/index/IndexLogManagerImplTest.scala]: done

File: src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala (hunk 12 above)
Comment (saucam) [reply src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala]: done

File: src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala (hunk 11 above)
Comment (saucam) [reply src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala]: ok done

File: src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala (hunk 10 above)
Comment (saucam) [reply src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala]: added

**Comment by saucam** [comment]
@imback82 sorry for so many style issues!! All fixed

**Comment by apoorvedave1** [comment]
LGTM, thanks @saucam

**Comment by saucam** [comment]
Thanks!!
